File: src/oodle/Oodle.java
/**
 * 
 */
package oodle;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.concurrent.TimeUnit;

import com.ochafik.lang.jnaerator.runtime.NativeSize;
import com.sun.jna.Memory;
import com.sun.jna.Pointer;

/**
 * @author FunGames
 *
 */
public class Oodle {
	private static OodleLibrary oodleLib = null;

	public static byte[] oodleDecompress(byte[] src, int dst_Length) {
		if (!loadLib())
			return null;
		
		int src_Length = src.length;
		byte[] dst = new byte[dst_Length];
		
		long start = System.currentTimeMillis();
		Memory sourcePointer = new Memory(src_Length);
		sourcePointer.write(0, src, 0, src_Length);
		Memory resultPointer = new Memory(dst_Length);
		resultPointer.write(0, dst, 0, dst_Length);
		int resultCode = oodleLib.OodleLZ_Decompress(new OodleLibrary.uint8(sourcePointer), src_Length, new OodleLibrary.uint8(resultPointer),  new NativeSize(dst_Length), 0, 0, 0, new OodleLibrary.uint8(), new NativeSize(0), com.sun.jna.Pointer.createConstant(0), com.sun.jna.Pointer.createConstant(0), com.sun.jna.Pointer.createConstant(0), new NativeSize(0), 0);
		double seconds = (double) ((System.currentTimeMillis() - start) / 1000) % 60 ;
		if(resultCode <= 0) {
			System.err.println("Oodle Decompression failed: " + resultCode);
			return null;
		}
		byte[] res = resultPointer.getByteArray(0, resultCode); 
		double mbPerSec = src_Length * 1e-6 / seconds;
		System.out.println(String.format("Oodle Decompress: %d => %d (%.2f seconds, %.2f MB/s)", src_Length, resultCode, seconds, mbPerSec));
		return res;
	}
	
	public static byte[] oodleCompress(byte[] input, int compressor, int compressionLevel) {
		if (!loadLib())
			return null;
		
		int input_Size = input.length;
		int dst_Length = input_Size + 65536;
		byte[] dst = new byte[dst_Length];
		
		long start = System.currentTimeMillis();
		Memory inputPointer = new Memory(input_Size);
		inputPointer.write(0, input, 0, input_Size);
		Memory resultPointer = new Memory(dst_Length);
		resultPointer.write(0, dst, 0, dst_Length);
		int resultCode = oodleLib.OodleLZ_Compress(compressor, new OodleLibrary.uint8(inputPointer), new NativeSize(input_Size), new OodleLibrary.uint8(resultPointer), compressionLevel, Pointer.createConstant(0), new NativeSize(0), new NativeSize(0), Pointer.createConstant(0), new NativeSize(0));
		double seconds = (double) ((System.currentTimeMillis() - start) / 1000) % 60 ;
		if(resultCode <= 0) {
			System.err.println("Oodle Compression failed: " + resultCode);
			return null;
		}
		byte[] res = resultPointer.getByteArray(0, resultCode);
		double mbPerSec = input_Size * 1e-6 / seconds;
		System.out.println(String.format("Oodle Compress: %d => %d (%.2f seconds, %.2f MB/s)", input_Size, resultCode, seconds, mbPerSec));
		return res;
	}

	private static boolean loadLib() {
		if (oodleLib != null)
			return true;
		System.setProperty("jna.library.path", System.getProperty("user.dir"));
		InputStream in = Oodle.class.getResourceAsStream("oo2core_7_win64.dll");

		FileOutputStream fos;
		try {
			fos = new FileOutputStream(new File(System.getProperty("user.dir") + "/oo2core_7_win64.dll"));
			int read = -1;
			byte[] buffer = new byte[1024];
			while ((read = in.read(buffer)) != -1) {
				fos.write(buffer, 0, read);
			}
			in.close();
			fos.close();
			oodleLib = OodleLibrary.INSTANCE;
			return oodleLib != null;
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return false;
	}
}

File: src/main/kotlin/me/fungames/oodle/Oodle.kt
package me.fungames.oodle

import com.sun.jna.Memory
import com.sun.jna.Native
import com.sun.jna.Pointer
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileOutputStream

const val COMPRESSOR_LZH = 0
const val COMPRESSOR_LZHLW = 1
const val COMPRESSOR_LZNIB = 2
const val COMPRESSOR_NONE = 3
const val COMPRESSOR_LZB16 = 4
const val COMPRESSOR_LZBLW = 5
const val COMPRESSOR_LZA = 6
const val COMPRESSOR_LZNA = 7
const val COMPRESSOR_KRAKEN = 8
const val COMPRESSOR_MERMAID = 9
const val COMPRESSOR_BITKNIT = 10
const val COMPRESSOR_SELKIE = 11
const val COMPRESSOR_HYDRA = 12
const val COMPRESSOR_LEVIATHAN = 13

const val COMPRESSION_LEVEL_NONE = 0
const val COMPRESSION_LEVEL_SUPER_FAST = 1
const val COMPRESSION_LEVEL_VERY_FAST = 2
const val COMPRESSION_LEVEL_FAST = 3
const val COMPRESSION_LEVEL_NORMAL = 4
const val COMPRESSION_LEVEL_OPTIMAL1 = 5
const val COMPRESSION_LEVEL_OPTIMAL2 = 6
const val COMPRESSION_LEVEL_OPTIMAL3 = 7
const val COMPRESSION_LEVEL_OPTIMAL4 = 8
const val COMPRESSION_LEVEL_OPTIMAL5 = 9

/**
 * Singleton for Oodle decompression and compression
 */
object Oodle {
    private val logger = LoggerFactory.getLogger("Oodle")
    private lateinit var oodleLib: OodleLibrary

    /**
     * Decompresses an Oodle compressed array
     * @param src the compressed source data
     * @param dstLen the uncompressed length
     * @return the decompressed data
     * @throws DecompressException when the decompression fails
     */
    @Throws(DecompressException::class)
    @JvmStatic
    fun decompress(src: ByteArray, dstLen: Int) = ByteArray(dstLen).also { decompress(src, it) }

    /**
     * Decompresses an Oodle compressed array
     * @param src the compressed source data
     * @param dst the destination buffer
     * @throws DecompressException when the decompression fails
     * @throws IllegalStateException when the library could not be loaded
     */
    @Throws(DecompressException::class)
    @JvmStatic
    fun decompress(src: ByteArray, dst: ByteArray) {
        decompress(src, 0, src.size, dst, 0, dst.size)
    }

    /**
     * Decompresses an Oodle compressed array
     * @param src the compressed source data
     * @param srcOff the offset into `src`
     * @param srcLen the compressed length
     * @param dst the destination buffer
     * @param dstOff the offset into `dst`
     * @param dstLen the uncompressed length
     * @throws DecompressException when the decompression fails
     * @throws IllegalStateException when the library could not be loaded
     */
    @Throws(DecompressException::class)
    @JvmStatic
    fun decompress(src: ByteArray, srcOff: Int, srcLen: Int, dst: ByteArray, dstOff: Int, dstLen: Int) {
        ensureLib()
        val start = System.currentTimeMillis()
        val sourcePointer = Memory(srcLen.toLong())
        sourcePointer.write(0L, src, srcOff, srcLen)
        val dstPointer = Memory(dstLen.toLong())
        val resultCode = oodleLib.OodleLZ_Decompress(
            sourcePointer, srcLen,
            dstPointer, dstLen.toLong(),
            0, 0, Integer.MAX_VALUE, Pointer.NULL, 0L, Pointer.NULL, Pointer.NULL, Pointer.NULL, 0L, 0
        )
        if (resultCode <= 0)
            throw DecompressException("Oodle decompression failed with code $resultCode")
        dstPointer.read(0, dst, dstOff, dstLen)
        val stop = System.currentTimeMillis()
        val seconds = (stop - start).toFloat() / 1000
        logger.trace("Oodle decompress: $srcLen => $dstLen ($seconds seconds)")
    }

    /**
     * Compresses a byte array
     * @param uncompressed the uncompressed source data
     * @param compressor the compressor to use
     * @param compressionLevel the compression level to use
     * @return the compressed data
     * @throws CompressException when the compression fails
     * @throws IllegalStateException when the library could not be loaded
     */
    @Throws(CompressException::class)
    @JvmStatic
    fun compress(uncompressed: ByteArray, compressor: Int, compressionLevel: Int): ByteArray {
        ensureLib()
        val start = System.currentTimeMillis()
        val srcLength = uncompressed.size
        val dstLength = srcLength + 65536
        val sourcePointer = Memory(srcLength.toLong())
        sourcePointer.write(0L, uncompressed, 0, srcLength)
        val dstPointer = Memory(dstLength.toLong())
        val resultCode = oodleLib.OodleLZ_Compress(
            compressor,
            sourcePointer, srcLength.toLong(),
            dstPointer, compressionLevel,
            Pointer.NULL, 0, 0, Pointer.NULL, 0
        )
        if (resultCode <= 0)
            throw CompressException("Oodle compression failed with code $resultCode")
        val dst = dstPointer.getByteArray(0, resultCode)
        val stop = System.currentTimeMillis()
        val seconds = (stop - start).toFloat() / 1000
        logger.trace("Oodle decompress: $srcLength => $dstLength ($seconds seconds)")
        return dst
    }

    @Throws(IllegalStateException::class)
    private fun ensureLib() {
        if (::oodleLib.isInitialized)
            return
        val oodleLibName = System.getProperty("JOodle.libName", "oo2core_7_win64.dll")
        val oodleLibFile = File(oodleLibName)
        var canLoad = oodleLibFile.exists()
        if (!canLoad && System.getProperty("os.name").contains("Windows")) {
            val input = this::class.java.getResourceAsStream("/oo2core_7_win64.dll")
            if (input != null) {
                val out = FileOutputStream(oodleLibFile)
                input.copyTo(out)
                input.close()
                out.close()
                canLoad = true
            }
        }
        if (!canLoad){
            throw IllegalStateException("Oodle library could not be loaded")
        }
        System.setProperty("jna.library.path", System.getProperty("user.dir"))
        try {
            oodleLib = Native.load(oodleLibName, OodleLibrary::class.java)
        } catch (e: Exception) {
            throw IllegalStateException("Oodle library failed to load", e)
        }
    }
}